apps/services/license-api/src/app/modules/license/clients/firearmLicense/firearmLicenseApiClient.service.ts

```diff
@@ -0,0 +1,165 @@
+import { Inject, Injectable } from '@nestjs/common'
+import { GenericLicenseClient } from '../../license.types'
+import { OpenFirearmApi } from '@island.is/clients/firearm-license'
+import type { Logger } from '@island.is/logging'
+import { LOGGER_PROVIDER } from '@island.is/logging'
+import {
+  SmartSolutionsApi,
+  PassDataInput,
+  Pass,
+  RevokePassData,
+  VerifyPassData,
+  Result,
+} from '@island.is/clients/smartsolutions'
+import { createPkPassDataInput } from './firearmLicenseMapper'
+import {
+  format as formatNationalId,
+  sanitize as sanitizeNationalId,
+} from 'kennitala'
+
+@Injectable()
+export class FirearmLicenseApiClientService implements GenericLicenseClient {
+  constructor(
+    @Inject(LOGGER_PROVIDER) private logger: Logger,
+    private firearmApi: OpenFirearmApi,
+    private smartApi: SmartSolutionsApi,
+  ) {}
+
+  pushUpdate(
+    inputData: PassDataInput,
+    nationalId: string,
+  ): Promise<Result<Pass | undefined>> {
+    return this.smartApi.updatePkPass(inputData, formatNationalId(nationalId))
+  }
+
+  async pullUpdate(nationalId: string): Promise<Result<Pass | undefined>> {
+    let data
+    try {
+      data = await Promise.all([
+        this.firearmApi.getVerificationLicenseInfo(nationalId),
+        this.firearmApi.getVerificationPropertyInfo(nationalId),
+      ])
+    } catch (e) {
+      return {
+        ok: false,
+        error: {
+          code: 13,
+          message: 'Service error',
+          data: JSON.stringify(e),
+        },
+      }
+    }
+
+    const [licenseInfo, propertyInfo] = data
+    if (!licenseInfo) {
+      return {
+        ok: false,
+        error: {
+          code: 3,
+          message: 'No license info found for user',
+        },
+      }
+    }
+
+    const inputValues = createPkPassDataInput(
+      licenseInfo,
+      propertyInfo,
+      nationalId,
+    )
+
+    if (!inputValues) {
+      return {
+        ok: false,
+        error: {
+          code: 4,
+          message: 'Mapping failed, invalid data',
+        },
+      }
+    }
+
+    const thumbnail = licenseInfo.licenseImgBase64
+    const payload: PassDataInput = {
+      inputFieldValues: inputValues,
+      thumbnail: thumbnail
+        ? {
+            imageBase64String: thumbnail
+              .substring(thumbnail.indexOf(',') + 1)
+              .trim(),
+          }
+        : null,
+    }
+
+    return this.smartApi.updatePkPass(payload, formatNationalId(nationalId))
+  }
+
+  revoke(nationalId: string): Promise<Result<RevokePassData>> {
+    return this.smartApi.revokePkPass(formatNationalId(nationalId))
+  }
+
+  /** We need to verify the pk pass AND the license itself! */
+  async verify(inputData: string): Promise<Result<VerifyPassData>> {
+    //need to parse the scanner data
+    const { code, date } = JSON.parse(inputData)
```

Comment:
Do we need to verify the output of the parse before moving forward? In particular `inputData = 'null'` will throw a TypeError and `inputData = ''` a SyntaxError.

Reply:
will implement better input validation